lambeq/training/cached_tn_path_optimizer.py

```diff
@@ -0,0 +1,385 @@
+# Copyright 2021-2025 Cambridge Quantum Computing Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Opt-Einsum contractor for reusing previously computed paths.
+"""
+
+import functools
+from pathlib import Path
+import pickle
+from typing import (
+    Collection, Iterable, Sequence, TypeVar
+)
+
+import opt_einsum as oe  # type: ignore[import-untyped]
+from tensornetwork import (
+    AbstractNode, contract_between, contract_parallel,
+    Edge, get_all_edges, get_subgraph_dangling
+)
+from tensornetwork.contractors.opt_einsum_paths import utils
+
+from lambeq.training.checkpoint import Checkpoint
+
+
+T = TypeVar('T')
+ContractionPath = Collection[tuple[int, ...]]
+ContractionKey = tuple[
+    tuple[tuple[str, ...], ...],
+    tuple[str, ...],
+    tuple[tuple[str, int], ...]
+]
+
+
+def _dedup(seq: Iterable[T]) -> list[T]:
+    """
+    Remove duplicates from list while maintaining order of the items.
+    Note: this method was taken from
+    https://stackoverflow.com/questions/480214/how-do-i-remove-duplicates-
+    from-a-list-while-preserving-order
+    """
+    seen: set[T] = set()
+    # Alias add() so python doesn't have to check it each time we call it
+    seen_add = seen.add
+    return [x for x in seq if not (x in seen or seen_add(x))]
+
+
+class TnPathOptimizer(oe.paths.PathOptimizer):
+    """Opt-einsum custom optimizer."""
+    _optimizer: functools.partial[ContractionPath]
+    memory_limit: int | None
+
+    def __init__(
+        self,
+        algorithm: str = 'auto',
+        memory_limit: int | None = None,
+        **kwargs
+    ):
+        """
+        parameters
```

Comment:
Let's capitalise properly first letters of headings and descriptions as in the rest of the code.